hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/HFileOutputFormat2.java

```diff
@@ -254,16 +255,19 @@ public void write(ImmutableBytesWritable row, V cell) throws IOException {
         byte[] tableNameBytes = null;
         if (writeMultipleTables) {
           tableNameBytes = MultiTableHFileOutputFormat.getTableName(row.get());
-          tableNameBytes = TableName.valueOf(tableNameBytes).toBytes();
+          tableNameBytes = TableName.valueOf(tableNameBytes).getNameWithNamespaceInclAsString()
```

Comment:
Applications (HFile readers) that read things written by HFileOutputFormat2 may need to change. There is a path building semantic change implied by `getNameWithNamespaceInclAsString` being necessary now for something introduced by this patch. Is this a compatibility problem? Are these changes necessary?

Reply:
@apurtell 
This is on condition multi table WALPlayer run. This condition seems to be true only for Backup from what I could understand. From that perspective seems alright to me --> [HBASE-20530](https://github.com/apache/hbase/commit/acbc3a225338fd1ff82226ebbd937f7b15ef5b60)

Reply:
I've been looking at this. I think the assumptions in HBASE-20530 are incorrect. The multi-table support in HFileOutputFormat2 was introduced separately from backups and already exists in branch-2. As such we can't break them like this. I think changing this path would fall under "File format compatibility" in our [compatibility rules](https://hbase.apache.org/book.html#hbase.versioning.post10) which requires a major version to change.

It seems like the problem in HBASE-20530 was due to a mismatch between the format returned by `HBackupFileSystem#getTableBackupDir` and this code here. The decision in that jira was to change this code here, rather than getTableBackupDir. I think that was a reasonable decision for two reasons:
1. it was a major version upgrade at the time.
2. this code here in HFileOutputFormat2 is susceptible to a bug where multiple tables with the same name but in separate namespaces might collide.

That bug seems problematic, but no one has reported it. We have two options:

1. Leave this code alone here, and instead change `HBackupFileSystem#getTableBackupDir`.
2. Add a config in HFileOutputFormat2 like `hbase.mapreduce.hfileoutputformat.multi.table.include.namespace`. We can have B&R always set this to true, and the config would be removed as part of 3.x upgrade. We'd use this to gate the logic above.

I don't love adding more config params, but I might suggest choosing option 2. This way if anyone stumbles across this bug in their own use-case, they actually have a way to get around it.

Reply:
I reviewed the rest of this patch -- the modified classes not the net-new ones (which i confirmed are identical to `master`).  This change here (and the related test changes) is the only concerning thing. I think once we fix this we should be able to move forward with merging. 

@apurtell to your question "is this change necessary?" -- Yes, _some_ change is necessary due to a mismatch between the output path here and the path expected in `HBackupFileSystem#getTableBackupDir`. I gave 2 options above to make this backwards compatible, do you have any opinion on them? 

I know we've discussed configurations in the past :) I'd like to avoid unnecessary configurations generally, but a new one here (option 2) might be best.

Reply:
@bbeaudreault  Thanks for the inputs. Let me think over and come back with my view on the options you suggested.

Reply:
@bbeaudreault  I am inclined towards option 2. `Add a config in HFileOutputFormat2 like`, adding more configs isn't particularly neat. It comes with a timeframe and that is 2.x only. Shall I go ahead in that route?

Reply:
Yes, let's go ahead with that. Thanks @rda3mon!